core/drivers/crypto/hisilicon/sec_cipher.c

```diff
@@ -0,0 +1,780 @@
+	if (tt)
+		data[0] ^= 0x87;
+}
+
+/*
+ * When the IV is delivered by segment,
+ * the AES/SM4-ECB is used to update the IV to be used next time.
+ */
+static TEE_Result xts_iv_update(struct sec_cipher_ctx *c_ctx)
+{
+	size_t xts_key_len = c_ctx->key_len / 2;
+	struct sec_cipher_ctx ecb_ctx = { 0 };
+	TEE_Result ret = TEE_SUCCESS;
+	size_t i = 0;
+
+	ecb_ctx.alg = c_ctx->alg;
+	ecb_ctx.mode = C_MODE_ECB;
+	ret = sec_cipher_set_key(&ecb_ctx, c_ctx->key + xts_key_len,
+				 xts_key_len, NULL, 0);
+	if (ret)
+		return ret;
+
+	ecb_ctx.encrypt = true;
+	ecb_ctx.in = (uint8_t *)c_ctx->iv;
+	ecb_ctx.out = (uint8_t *)c_ctx->iv;
+	ecb_ctx.in_dma = c_ctx->iv_dma;
+	ecb_ctx.out_dma = c_ctx->iv_dma;
+	ecb_ctx.len = c_ctx->iv_len;
+
+	ret = sec_do_cipher_task(c_ctx->qp, &ecb_ctx);
+	if (ret) {
+		EMSG("Fail to encrypt xts iv, ret=%#"PRIx32, ret);
+		return ret;
+	}
+
+	for (i = 0; i < DIV_ROUND_UP(c_ctx->len, AES_SM4_BLOCK_SIZE); i++)
+		xts_multi_galois((uint8_t *)c_ctx->iv);
+
+	ecb_ctx.encrypt = false;
+	ret = sec_do_cipher_task(c_ctx->qp, &ecb_ctx);
+	if (ret)
+		EMSG("Fail to decrypt xts iv, ret=%#"PRIx32, ret);
+
+	return ret;
+}
+
+static TEE_Result sec_update_iv(struct sec_cipher_ctx *c_ctx)
+{
+	TEE_Result ret = TEE_SUCCESS;
+	size_t offset = 0;
+
+	switch (c_ctx->mode) {
+	case C_MODE_CBC:
+		offset = c_ctx->len - c_ctx->iv_len;
+		if (c_ctx->encrypt && c_ctx->len >= c_ctx->iv_len)
+			memcpy(c_ctx->iv, c_ctx->out + offset, c_ctx->iv_len);
+		if (!c_ctx->encrypt && c_ctx->len >= c_ctx->iv_len)
+			memcpy(c_ctx->iv, c_ctx->in + offset, c_ctx->iv_len);
+		break;
+	case C_MODE_CTR:
+		/*
+		 * Increase the iv counter with the number of processed blocks.
+		 */
+		ctr_iv_inc(c_ctx->iv, c_ctx->len >> CTR_MODE_LEN_SHIFT);
+		break;
+	case C_MODE_XTS:
+		ret = xts_iv_update(c_ctx);
+		break;
+	default:
+		break;
+	}
+
+	return ret;
+}
+
+static TEE_Result sec_cipher_iv_check(struct sec_cipher_ctx *c_ctx,
+				      size_t iv_size)
+{
+	TEE_Result ret = TEE_ERROR_BAD_PARAMETERS;
+
+	switch (c_ctx->mode) {
+	case C_MODE_ECB:
+		if (!iv_size)
+			ret = TEE_SUCCESS;
+		break;
+	case C_MODE_CBC:
+		if (c_ctx->alg == C_ALG_DES || c_ctx->alg == C_ALG_3DES) {
+			if (iv_size == DES_CBC_IV_SIZE)
+				ret = TEE_SUCCESS;
+			break;
+		}
+		fallthrough;
+	case C_MODE_XTS:
+	case C_MODE_CTR:
+		if (c_ctx->alg == C_ALG_AES || c_ctx->alg == C_ALG_SM4) {
+			if (iv_size == AES_SM4_IV_SIZE)
+				ret = TEE_SUCCESS;
+		}
+		break;
+	default:
+		break;
+	}
+
+	if (ret)
+		EMSG("Fail to check iv_size");
+
+	return ret;
+}
+
+static TEE_Result sec_cipher_set_iv(struct sec_cipher_ctx *c_ctx,
+				    uint8_t *iv, size_t iv_len)
+{
+	TEE_Result ret = TEE_SUCCESS;
+
+	if (!iv && iv_len) {
+		EMSG("iv is NULL");
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	ret = sec_cipher_iv_check(c_ctx, iv_len);
+	if (ret)
+		return ret;
+
+	c_ctx->iv_len = iv_len;
+	c_ctx->iv_dma = virt_to_phys(c_ctx->iv);
+	if (!c_ctx->iv_dma) {
+		EMSG("c_iv_dma is NULL");
+		return TEE_ERROR_OUT_OF_MEMORY;
```

Comment:
Should rather be `TEE_ERROR_BAD_PARAMETERS` or something else.
`TEE_ERROR_OUT_OF_MEMORY` relates to memory allocation issues which is not the case here.